cpp/src/arrow/filesystem/azurefs_test.cc

```diff
@@ -71,56 +72,113 @@ using ::testing::Not;
+      : account_name_(std::move(account_name)), account_key_(std::move(account_key)) {}
 
-class AzuriteEnv : public ::testing::Environment {
  public:
-  AzuriteEnv() {
-    account_name_ = "devstoreaccount1";
-    account_key_ =
-        "Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/"
-        "KBHBeksoGMGw==";
-    auto exe_path = bp::search_path("azurite");
-    if (exe_path.empty()) {
-      auto error = std::string("Could not find Azurite emulator.");
-      status_ = Status::Invalid(error);
-      return;
+  ~BaseAzureEnv() override = default;
+
+  const std::string& account_name() const { return account_name_; }
+  const std::string& account_key() const { return account_key_; }
+
+  virtual AzureBackend backend() const = 0;
+
+  virtual bool WithHierarchicalNamespace() const { return false; }
+
+  virtual Result<int64_t> GetDebugLogSize() { return 0; }
+  virtual Status DumpDebugLog(int64_t position) {
+    return Status::NotImplemented("BaseAzureEnv::DumpDebugLog");
+  }
+};
+
+template <class AzureEnvClass>
+class AzureEnvImpl : public BaseAzureEnv {
+ protected:
+  static Result<BaseAzureEnv*> MakeAndAddToGlobalTestEnvironment() {
+    ARROW_ASSIGN_OR_RAISE(auto instance, AzureEnvClass::Make());
+    auto* heap_ptr = instance.release();
+    ::testing::AddGlobalTestEnvironment(heap_ptr);
+    return heap_ptr;
+  }
+
+  using BaseAzureEnv::BaseAzureEnv;
+
+  static Result<std::unique_ptr<AzureEnvClass>> MakeFromEnvVars(
+      const std::string& account_name_var, const std::string& account_key_var) {
+    const auto account_name = std::getenv(account_name_var.c_str());
+    const auto account_key = std::getenv(account_key_var.c_str());
+    if (!account_name) {
+      return Status::Cancelled(account_name_var + " not set.");
     }
-    auto temp_dir_ = *TemporaryDir::Make("azurefs-test-");
-    auto debug_log_path_result = temp_dir_->path().Join("debug.log");
-    if (!debug_log_path_result.ok()) {
-      status_ = debug_log_path_result.status();
-      return;
+    if (!account_key) {
+      return Status::Cancelled(account_key_var + " not set.");
     }
-    debug_log_path_ = *debug_log_path_result;
-    server_process_ =
-        bp::child(boost::this_process::environment(), exe_path, "--silent", "--location",
-                  temp_dir_->path().ToString(), "--debug", debug_log_path_.ToString());
-    if (!(server_process_.valid() && server_process_.running())) {
-      auto error = "Could not start Azurite emulator.";
-      server_process_.terminate();
-      server_process_.wait();
-      status_ = Status::Invalid(error);
-      return;
+    return std::unique_ptr<AzureEnvClass>{new AzureEnvClass(account_name, account_key)};
+  }
+
+ public:
+  ~AzureEnvImpl() override = default;
+
+  static Result<BaseAzureEnv*> GetInstance() {
+    static auto env = MakeAndAddToGlobalTestEnvironment();
+    if (env.ok()) {
+      return env;
     }
-    status_ = Status::OK();
+    return env.status();
```

Comment:
Can we just always return `env`?
(`return env`)

Reply:
Haha. Yes. I don't know how I ended up on this.